build-system/compile/pre-closure-babel.js

```diff
@@ -52,69 +60,47 @@ function getFilesToTransform() {
 }
 
 /**
- * Apply babel transforms prior to closure compiler pass.
+ * Apply babel transforms prior to closure compiler pass and return the path
+ * of the transformed file.
  *
- * When a source file is transformed for the first time, it is written to an
- * in-memory cache from where it is retrieved every subsequent time without
+ * When a source file is transformed for the first time, it is written to a
+ * cache directory from where it is retrieved every subsequent time without
  * invoking babel.
  *
- * @return {!Promise}
+ * @param {string} file
+ * @param {string} outputFilename
+ * @param {!Object} options
+ * @return {string}
  */
-function preClosureBabel() {
-  const babel = gulpBabel({caller: {name: 'pre-closure'}});
-
-  return through.obj((file, enc, next) => {
-    if (!filesToTransform) {
-      filesToTransform = getFilesToTransform();
-    }
-    if (!filesToTransform.includes(file.relative)) {
-      return next(null, file);
-    }
-
-    if (cache.has(file.path)) {
-      return next(null, cache.get(file.path));
-    }
-
-    let data, err;
-    debug(
-      CompilationLifecycles['pre-babel'],
-      file.path,
-      file.contents,
-      file.sourceMap
-    );
-    /**
-     * @param {*} d
-     * @return {void}
-     */
-    function onData(d) {
-      babel.off('error', onError);
-      data = d;
-    }
-    /**
-     * @param {Error} e
-     * @return {void}
-     */
-    function onError(e) {
-      babel.off('data', onData);
-      err = e;
-    }
-    babel.once('data', onData);
-    babel.once('error', onError);
-    babel.write(file, enc, () => {
-      if (err) {
-        return next(err);
-      }
-
-      debug(
-        CompilationLifecycles['pre-closure'],
-        file.path,
-        data.contents,
-        data.sourceMap
-      );
-      cache.set(file.path, data);
-      next(null, data);
-    });
+async function preClosureBabel(file, outputFilename, options) {
+  if (!cacheDir) {
+    cacheDir = tempy.directory();
+  }
+  if (!filesToTransform) {
+    filesToTransform = getFilesToTransform();
+  }
+  const transformedFile = path.join(cacheDir, file);
+  if (fs.existsSync(transformedFile)) {
+    return transformedFile;
+  }
+  if (!filesToTransform.includes(file)) {
+    await fs.copy(file, transformedFile);
+    return transformedFile;
+  }
+  debug(CompilationLifecycles['pre-babel'], file);
+  const babelOptions = babel.loadOptions({
+    caller: {name: 'pre-closure'},
+    filename: file,
+    sourceFileName: path.relative(process.cwd(), file),
   });
+  try {
+    const {code} = await babel.transformFileAsync(file, babelOptions);
+    await fs.outputFile(transformedFile, Buffer.from(code, 'utf-8'));
+    debug(CompilationLifecycles['pre-closure'], transformedFile);
+  } catch (err) {
+    handlePreClosureError(err, outputFilename, options);
```

Comment:
Similar question here as earlier: in the case of `options.continueOnError`, I don't see any `throw` or `reject` or `resolve` being called, so the parent Promise will resolve with `transformedFile`. Thats intended here though right?

Reply:
I've changed this as well. The function now returns an `Error` if needed, and the caller explicitly throws it.